File: Algorithms/Fibonacci.py
"""
This contains two implementations of Fibonacci Sequence by using recursivity
The first implementation is a bad one because it uses recursivity in a bad way
by computing the same values many times. The complexity of this implementation
is exponential: O(2^n)
The second implementation uses recurcive calls in a smart way to avoid redondancies.
The complexity of this implementation is linear: O(n).
"""
def bad_fibonacci(n):
    #Return the nth Fibonacci number.
    if n <=1:
        return n
    else:
        return bad_fibonacci(n−2) + bad_fibonacci(n−1)

def good_fibonacci(n):
    #Return pair of Fibonacci numbers, F(n) and F(n-1).
    if n <=1:
        return (n,0)
    else:
        (a, b) = good_fibonacci(n−1)
        return (a+b, a)

File: Algorithms/Javascript/fibonacci.js
/**
 * Fibonacci
 * calculating value of n(th) place in Fibonacci sequence
 * f(n) = f(n-2) + f(n-1)
 * f(0) = 0
 * f(1) = 1
 * @param {*} n
 * @returns
 */
function fibonacci(n) {
  if (n == 0 || n == 1) {
    return n;
  } else {
    return fibonacci(n - 2) + fibonacci(n - 1);
  }
}
